File: abstract_factory/abstract_factory.go
package main

import "fmt"

// this is my concrete girl friend
type GirlFriend struct {
	nationality string
	eyesColor   string
	language    string
}

type Baby struct {
	nationality string
}

type AbstractFactory interface {
	CreateMyLove() GirlFriend
	CreateBaby() Baby
}

type IndianFactory struct{}

type KoreanFactory struct{}

func (a IndianFactory) CreateMyLove() GirlFriend {
	return GirlFriend{"Indian", "Black", "Hindi"}
}

func (a IndianFactory) CreateBaby() Baby {
	return Baby{"Hindi"}
}

func (a KoreanFactory) CreateMyLove() GirlFriend {
	return GirlFriend{"Korean", "Brown", "Korean"}
}

func (a KoreanFactory) CreateBaby() Baby {
	return Baby{"Korean"}
}

func getFacotry(typeGf string) interface{} {
	factoryMap := map[string]AbstractFactory{
		"Indian": IndianFactory{},
		"Korean": KoreanFactory{},
	}
	return factoryMap[typeGf]
}

func main() {
	var factory interface{}
	factory = getFacotry("Indian")
	fmt.Println(factory.(IndianFactory).CreateMyLove().eyesColor)
	fmt.Println(factory.(IndianFactory).CreateBaby().nationality)
	factory = getFacotry("Korean")
	fmt.Println(factory.(KoreanFactory).CreateMyLove().eyesColor)
	fmt.Println(factory.(KoreanFactory).CreateBaby().nationality)
}

File: abstract_factory/abstract_factory.py
import abc


class GirlFriend(object):

    def __init__(self, nationality, eyesColor, language):
        self.nationality = nationality
        self.eyesColor = eyesColor
        self.language = language


class Baby(object):

    def __init__(self, nationality):
        self.nationality = nationality


class AbstractFactory(object):
    """
    Declare an interface for operations that create abstract product
    objects.
    """
    __metaclass__ = abc.ABCMeta

    @abc.abstractmethod
    def create_myLove(self):
        pass

    @abc.abstractmethod
    def create_baby(self):
        pass


class IndianGirlFriendFactory(AbstractFactory):

    nationality = "Indian"
    eyesColor = "black"
    language = "Hindi"

    def create_myLove(self):
        return GirlFriend(self.nationality, self.eyesColor, self.language)

    def create_baby(self):
        return Baby(self.nationality)


class KoreanGirlFriendFactory(AbstractFactory):

    nationality = "Korean"
    eyesColor = "brown"
    language = "Korean"

    def create_myLove(self):
        return GirlFriend(self.nationality, self.eyesColor, self.language)

    def create_baby(self):
        return Baby(self.nationality)


factory = IndianGirlFriendFactory()
print(factory.create_myLove().eyesColor)
print(factory.create_baby().nationality)
factory = KoreanGirlFriendFactory()
print(factory.create_myLove().eyesColor)
print(factory.create_baby().nationality)

